channel_test.go

```diff
@@ -0,0 +1,181 @@
+// truncated if the msize would be exceeded.
+func TestWriteOverflow(t *testing.T) {
+	const (
+		msize         = 512
+		overflowMSize = msize * 3 / 2
+	)
+
+	var (
+		ctx   = context.Background()
+		conn  = &mockConn{}
+		ch    = NewChannel(conn, msize)
+		data  = bytes.Repeat([]byte{'A'}, overflowMSize)
+		fcall = newFcall(1, MessageTwrite{
+			Data: data,
+		})
+		messageSize uint32
+	)
+
+	if err := ch.WriteFcall(ctx, fcall); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := binary.Read(bytes.NewReader(conn.buf.Bytes()), binary.LittleEndian, &messageSize); err != nil {
+		t.Fatal(err)
+	}
+
+	if messageSize != msize {
+		t.Fatalf("should have truncated size header: %d != %d", messageSize, msize)
+	}
+
+	if conn.buf.Len() != msize {
+		t.Fatalf("should have truncated message: conn.buf.Len(%v) != msize(%v)", conn.buf.Len(), msize)
+	}
+}
+
+// TestWriteOverflowError ensures that we return an error in cases when there
+// will certainly be an overflow and it cannot be resolved.
+func TestWriteOverflowError(t *testing.T) {
+	const (
+		msize         = 4
+		overflowMSize = msize + 1
+	)
+
+	var (
+		ctx   = context.Background()
+		conn  = &mockConn{}
+		ch    = NewChannel(conn, msize)
+		data  = bytes.Repeat([]byte{'A'}, 4)
+		fcall = newFcall(1, MessageTwrite{
+			Data: data,
+		})
+		messageSize = 4 + ch.(*channel).codec.Size(fcall)
+	)
+
+	err := ch.WriteFcall(ctx, fcall)
+	if err == nil {
+		t.Fatal("error expected when overflowing message")
+	}
+
+	if Overflow(err) != messageSize-msize {
+		t.Fatalf("overflow should reflect messageSize and msize, %d != %d", Overflow(err), messageSize-msize)
+	}
+}
+
+// TestReadOverflow ensures that messages coming over a network connection do
+// not overflow the msize. Invalid messages will cause `ReadFcall` to return an
+// Overflow error.
+func TestReadOverflow(t *testing.T) {
+	const (
+		msize         = 256
+		overflowMSize = msize + 1
+	)
+
+	var (
+		ctx   = context.Background()
+		conn  = &mockConn{}
+		ch    = NewChannel(conn, msize)
+		data  = bytes.Repeat([]byte{'A'}, overflowMSize)
+		fcall = newFcall(1, MessageTwrite{
+			Data: data,
+		})
+		messageSize = 4 + ch.(*channel).codec.Size(fcall)
+	)
+
+	// prepare the raw message
+	p, err := ch.(*channel).codec.Marshal(fcall)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// "send" the message into the buffer
+	// this message is crafted to overflow the read buffer.
+	if err := sendmsg(&conn.buf, p); err != nil {
+		t.Fatal(err)
+	}
+
+	var incoming Fcall
+	err = ch.ReadFcall(ctx, &incoming)
+	if err == nil {
+		t.Fatal("expected error on fcall")
+	}
+
+	if Overflow(err) != messageSize-msize {
+		t.Fatalf("unexpected overflow on error: %v !=%v", Overflow(err), messageSize-msize)
+	}
+}
+
+// TestTreadRewrite ensures that messages that whose response would overflow
+// the msize will have be adjusted before sending.
+func TestTreadRewrite(t *testing.T) {
+	const (
+		msize         = 256
+		overflowMSize = msize + 1
```

Comment:
Same as with writes : I'd like to cover the case where len(p) fits in msize, but len(p) + [size of envelope] don't to the regression suite. That would avoid some potential naïve (and wrong) optimizations.